src/util.ts

```diff
@@ -953,3 +975,20 @@ export function isTokenDetectionEnabledForNetwork(chainId: string): boolean {
     chainId,
   );
 }
+
+/**
+ * Utility method to log if error is a common fetch error and otherwise rethrow it.
+ *
+ * @param error - Caught error that we should either rethrow or log to console.
+ */
+export function logOrRethrowError(error: any) {
+  if (
+    error.message?.match(/Fetch failed with status/u) ||
```

Comment:
This would flag for any unsuccessful fetch. e.g. if we submitted malformed parameters to the API and got back a 4XX error, this would catch that case. But we'd want that case to bubble up instead so that we get a Sentry report explaining that our code is broken.

Reply:
Ok. So do you think its okay to catch on `403`s in this method or perhaps pass an arg to `logOrRethrowError` with a platform specific error code we want to catch (`1020` in this case for instance)?

Reply:
Coming back to the general advice given earlier, what we want is to ensure _expected errors_ are handled in an appropriate way for the situation, and that _unexpected errors_ are thrown so that we can learn about this unanticipated situation and determine the best way to handle it. Does that make sense?

Right now with this function, any non-OK response from the server is being treated as expected and logged to the console. You've correctly pointed out that we should anticipate the API key being unset or invalid some of the time (e.g. for development builds, or in case our key gets revoked). But I would not anticipate us getting any other 4xx response status codes apart from 403, not unless we made a mistake somewhere.

I don't know what 1020 is supposed to indicate, so I don't know whether we should encode 403 overall as being expected, or 1020 specifically. Maybe if they have API documentation guaranteeing that a missing or invalid API key always returns that code, we could encode that expectation. Otherwise, encoding that might risk this breaking in prod if they change the API to return a different code.

Reply:
Probably best to just go by the 403. They confirmed they don't have docs for the 1020 error response, that it's just coming from cloudflare. So that might change in the future if they update how the API key check works or migrate to a new platform or something.

Reply:
@Gudahtt I've pursued this approach here: https://github.com/MetaMask/controllers/pull/805/commits/9a0d739752b0028765e42340895610ee0d665886

Let me know what you think

Reply:
@Gudahtt ping here when you have a moment